apps/configutl.c

```diff
@@ -0,0 +1,205 @@
+        }
+    }
+}
+
+/**
+ * Print all values in in the configuration section identified by section_name to stdout.
+ */
+static void print_section(BIO *out, const CONF *cnf, OPENSSL_CSTRING section_name)
+{
+    STACK_OF(CONF_VALUE) *values = NCONF_get_section(cnf, section_name);
+    int idx;
+
+    for (idx = 0; idx < sk_CONF_VALUE_num(values); idx++) {
+        CONF_VALUE *value = sk_CONF_VALUE_value(values, idx);
+
+        BIO_printf(out, "%s = ", value->name);
+        print_escaped_value(out, value->value);
+        BIO_printf(out, "\n");
+    }
+}
+
+typedef enum OPTION_choice {
+    OPT_COMMON,
+    OPT_DUMP,
+    OPT_OUT,
+    OPT_NOHEADER,
+    OPT_CONFIG
+} OPTION_CHOICE;
+
+const OPTIONS configutl_options[] = {
+    OPT_SECTION("General"),
+    {"help", OPT_HELP, '-', "Display this summary"},
+    {"dump", OPT_DUMP, '-', "Dump the linearized config file"},
+    {"config", OPT_CONFIG, 's', "Config file to deal with (the default one if omitted)"},
+    OPT_SECTION("Output"),
+    {"out", OPT_OUT, '>', "Output to filename rather than stdout"},
+    {"noheader", OPT_NOHEADER, '-', "Don't print the information about original config"},
+    {NULL}
+};
+
+/**
+ * Parse the passed OpenSSL configuration file (or the default one/specified in the
+ * OPENSSL_CONF environment variable) and write it back to stdout in
+ * a canonical format with all includes and variables expanded.
+ */
+int configutl_main(int argc, char *argv[])
+{
+    int ret = 1;
+    char *prog, *configfile = NULL;
+    OPTION_CHOICE o;
+    int dump = 0;
+    CONF *cnf = NULL;
+    long eline = 0;
+    int default_section_idx, idx;
+    int no_header = 0;
+    STACK_OF(OPENSSL_CSTRING) *sections = NULL;
+    BIO *out = NULL;
+    const char *outfile = NULL;
+
+    prog = opt_init(argc, argv, configutl_options);
+    while ((o = opt_next()) != OPT_EOF) {
+        switch (o) {
+        case OPT_EOF:
+        case OPT_ERR:
+ opthelp:
+            BIO_printf(bio_err, "%s: Use -help for summary.\n", prog);
+            goto end;
+        case OPT_HELP:
+            opt_help(configutl_options);
+            ret = 0;
+            goto end;
+            break;
+        case OPT_DUMP:
+            dump = 1;
+            break;
+        case OPT_NOHEADER:
+            no_header = 1;
+            break;
+        case OPT_CONFIG:
+            configfile = OPENSSL_strdup(opt_arg());
+            break;
+        case OPT_OUT:
+            outfile = opt_arg();
+            break;
+        }
+    }
+
+    if (dump == 0)
+        goto opthelp;
+
+    out = bio_open_default(outfile, 'w', FORMAT_TEXT);
+    if (out == NULL)
+        goto end;
+
+    if (configfile == NULL)
+        configfile = CONF_get1_default_config_file();
+
+    if (configfile == NULL)
+        goto end;
+
+    if ((cnf = NCONF_new(NULL)) == NULL)
+        goto end;
+
+    if (NCONF_load(cnf, configfile, &eline) == 0) {
+        BIO_printf(bio_err, "Error on line %ld of configuration file\n", eline);
```

Comment:
As far as I can tell from just code inspection, line numbers start at 0 (i.e. the first line in the file is line 0). Which I find slightly confusing (I would expect it to be 1). Not sure if that should be considered a bug in `NCONF_load`, or a documentation issue. Either way - should we add 1 to the line number here???

Also in the event of a problem loading the file, `eline` will be unchanged. We might want to initialise it to -1, and then test for that here to detect and report errors that have occurred before we've even opened the file.

Reply:
Done